akka-cluster-sharding/src/main/scala/akka/cluster/sharding/dynamic/DynamicShardAllocationStrategy.scala

```diff
@@ -0,0 +1,216 @@
+    // local only ask
+    implicit val timeout: Timeout = Timeout(5.seconds))
+    extends ShardCoordinator.ShardAllocationStrategy {
+
+  import DynamicShardAllocationStrategy._
+  import akka.pattern.ask
+  import system.dispatcher
+
+  private val log = Logging(system, classOf[DynamicShardAllocationStrategy])
+
+  private val shardState = createShardStateActor()
+
+  private[akka] def createShardStateActor(): ActorRef = {
+    system
+      .asInstanceOf[ExtendedActorSystem]
+      .systemActorOf(DDataStateActor.props(typeName), s"dynamic-allocation-state-$typeName")
+  }
+
+  private val cluster = Cluster(system)
+
+  override def allocateShard(
+      requester: ShardRegion,
+      shardId: ShardId,
+      currentShardAllocations: Map[ShardRegion, immutable.IndexedSeq[ShardId]]): Future[ShardRegion] = {
+
+    log.debug("allocateShard {} {} {}", shardId, requester, currentShardAllocations)
+
+    // current shard allocations include all current shard regions
+    (shardState ? GetShardLocation(shardId))
+      .mapTo[GetShardLocationResponse]
+      .map {
+        case GetShardLocationResponse(None) =>
+          log.debug("No specific location for shard {}. Allocating to requester {}", shardId, requester)
+          requester
+        case GetShardLocationResponse(Some(address)) =>
+          // if it is the local address, convert it so it is found in the shards
+          if (address == cluster.selfAddress) {
+            currentShardAllocations.keys.find(_.path.address.hasLocalScope) match {
+              case None =>
+                log.warning("unable to find local shard in currentShardAllocation. Using requester")
+                requester
+              case Some(localShardRegion) =>
+                log.debug("allocating to local shard")
+                localShardRegion
+            }
+          } else {
+            currentShardAllocations.keys.find(_.path.address == address) match {
+              case None =>
+                log.warning(
+                  "Dynamic shard location [{}] for shard {} not found in members [{}]",
+                  address,
+                  shardId,
+                  currentShardAllocations.keys.mkString(","))
+                requester
+              case Some(location) =>
+                log.debug("Allocating shard to location {}", location)
+                location
+            }
+          }
+      }
+      .recover {
+        case _: AskTimeoutException =>
+          log.warning("allocate timed out waiting for shard allocation state")
+          requester
+      }
+
+  }
+
+  override def rebalance(
+      currentShardAllocations: Map[ActorRef, immutable.IndexedSeq[ShardId]],
+      rebalanceInProgress: Set[ShardId]): Future[Set[ShardId]] = {
+
+    log.debug("rebalance {} {}", currentShardAllocations, rebalanceInProgress)
+
+    val currentAllocationByAddress: Map[Address, immutable.IndexedSeq[ShardId]] = currentShardAllocations.map {
+      case (_: LocalActorRef, value) =>
```

Comment:
maybe better to `ref: ActorRefScope if ref.isLocal`

Reply:
Thanks, hadn't come across that before